In [1]:
from carbon import CarbonSimulatorUI, __version__, __date__
print(f"Carbon Version v{__version__} ({__date__})", )

Carbon Version v2.2-BETA1 (02/Jan/2022)


# Carbon Simulation - Demo 4-2

In this demo we investigate the **Alpha Router performance** we add dust limit orders at either end of the main strats

Initialize an exact simulator and an alpha simulator

In [2]:
Sim = CarbonSimulatorUI(pair="ETH/USDC", verbose=False)
AlphaSim = CarbonSimulatorUI(pair="ETH/USDC", verbose=False, matching_method=CarbonSimulatorUI.MATCH_ALPHA)

Add identical strategies to both simulators

In [3]:
strat = ['ETH', 10, 2000, 2500, 1000, 2800, 2700]
for _ in range(10):
    Sim.add_strategy(*strat)
    AlphaSim.add_strategy(*strat)

for _ in range(3):
    Sim.add_strategy('ETH', 0.1, 1950, 1950, 1, 2750, 2750)
    AlphaSim.add_strategy('ETH', 0.1, 1950, 1950, 1, 2750, 2750)

#Sim.state()['orders']
AlphaSim.state()['orders']

id     pair   tkn   y_int       y y_unit  disabled  p_start   p_end  \
0    0  ETHUSDC   ETH    10.0    10.0    ETH     False   2000.0  2500.0   
1    1  ETHUSDC  USDC  1000.0  1000.0   USDC     False   2800.0  2700.0   
2    2  ETHUSDC   ETH    10.0    10.0    ETH     False   2000.0  2500.0   
3    3  ETHUSDC  USDC  1000.0  1000.0   USDC     False   2800.0  2700.0   
4    4  ETHUSDC   ETH    10.0    10.0    ETH     False   2000.0  2500.0   
5    5  ETHUSDC  USDC  1000.0  1000.0   USDC     False   2800.0  2700.0   
6    6  ETHUSDC   ETH    10.0    10.0    ETH     False   2000.0  2500.0   
7    7  ETHUSDC  USDC  1000.0  1000.0   USDC     False   2800.0  2700.0   
8    8  ETHUSDC   ETH    10.0    10.0    ETH     False   2000.0  2500.0   
9    9  ETHUSDC  USDC  1000.0  1000.0   USDC     False   2800.0  2700.0   
10  10  ETHUSDC   ETH    10.0    10.0    ETH     False   2000.0  2500.0   
11  11  ETHUSDC  USDC  1000.0  1000.0   USDC     False   2800.0  2700.0   
12  12  ETHUSDC   ETH    10.0    10.0    ETH     False   2000.0  2500.0   
13  13  ETHUSDC  USDC  1000.0  1000.0   USDC     False   2800.0  2700.0   
14  14  ETHUSDC   ETH    10.0    10.0    ETH     False   2000.0  2500.0   
15  15  ETHUSDC  USDC  1000.0  1000.0   USDC     False   2800.0  2700.0   
16  16  ETHUSDC   ETH    10.0    10.0    ETH     False   2000.0  2500.0   
17  17  ETHUSDC  USDC  1000.0  1000.0   USDC     False   2800.0  2700.0   
18  18  ETHUSDC   ETH    10.0    10.0    ETH     False   2000.0  2500.0   
19  19  ETHUSDC  USDC  1000.0  1000.0   USDC     False   2800.0  2700.0   
20  20  ETHUSDC   ETH     0.1     0.1    ETH     False   1950.0  1950.0   
21  21  ETHUSDC  USDC     1.0     1.0   USDC     False   2750.0  2750.0   
22  22  ETHUSDC   ETH     0.1     0.1    ETH     False   1950.0  1950.0   
23  23  ETHUSDC  USDC     1.0     1.0   USDC     False   2750.0  2750.0   
24  24  ETHUSDC   ETH     0.1     0.1    ETH     False   1950.0  1950.0   
25  25  ETHUSDC  USDC     1.0     1.0   USDC     False   2750.0  2750.0   

    p_marg        p_unit  lid  
0   2000.0  USDC per ETH    1  
1   2800.0  USDC per ETH    0  
2   2000.0  USDC per ETH    3  
3   2800.0  USDC per ETH    2  
4   2000.0  USDC per ETH    5  
5   2800.0  USDC per ETH    4  
6   2000.0  USDC per ETH    7  
7   2800.0  USDC per ETH    6  
8   2000.0  USDC per ETH    9  
9   2800.0  USDC per ETH    8  
10  2000.0  USDC per ETH   11  
11  2800.0  USDC per ETH   10  
12  2000.0  USDC per ETH   13  
13  2800.0  USDC per ETH   12  
14  2000.0  USDC per ETH   15  
15  2800.0  USDC per ETH   14  
16  2000.0  USDC per ETH   17  
17  2800.0  USDC per ETH   16  
18  2000.0  USDC per ETH   19  
19  2800.0  USDC per ETH   18  
20  1950.0  USDC per ETH   21  
21  2750.0  USDC per ETH   20  
22  1950.0  USDC per ETH   23  
23  2750.0  USDC per ETH   22  
24  1950.0  USDC per ETH   25  
25  2750.0  USDC per ETH   24

In [4]:
# set test amounts
usdc_trade_amount = 1000
eth_trade_amount = 0.5

## Route by Source

#### AMM buys USDC

First we test the standard simulator which should route through all 5 positions

In [5]:
Sim.amm_buys('USDC',usdc_trade_amount, execute=False)['trades']  # route_trade_by_source

uid  id subid                          note   aggr   exec limitfail  \
0   0.0   0     0                     route #20  False  False      None   
0   0.1   0     1                     route #22  False  False      None   
0   0.2   0     2                     route #24  False  False      None   
0   0.3   0     3                      route #0  False  False      None   
0   0.4   0     4                      route #2  False  False      None   
0   0.5   0     5                      route #4  False  False      None   
0   0.6   0     6                      route #6  False  False      None   
0   0.7   0     7                      route #8  False  False      None   
0   0.8   0     8                     route #10  False  False      None   
0   0.9   0     9                     route #12  False  False      None   
0  0.10   0    10                     route #14  False  False      None   
0  0.11   0    11                     route #16  False  False      None   
0  0.12   0    12                     route #18  False  False      None   
0     0   0     A  AMM sells 1ETH buys 1000USDC   True  False      None   

       amt1 tkn1    amt2  tkn2     pair  \
0  0.100000  ETH   195.0  USDC  ETHUSDC   
0  0.100000  ETH   195.0  USDC  ETHUSDC   
0  0.100000  ETH   195.0  USDC  ETHUSDC   
0  0.020745  ETH    41.5  USDC  ETHUSDC   
0  0.020745  ETH    41.5  USDC  ETHUSDC   
0  0.020745  ETH    41.5  USDC  ETHUSDC   
0  0.020745  ETH    41.5  USDC  ETHUSDC   
0  0.020745  ETH    41.5  USDC  ETHUSDC   
0  0.020745  ETH    41.5  USDC  ETHUSDC   
0  0.020745  ETH    41.5  USDC  ETHUSDC   
0  0.020745  ETH    41.5  USDC  ETHUSDC   
0  0.020745  ETH    41.5  USDC  ETHUSDC   
0  0.020745  ETH    41.5  USDC  ETHUSDC   
0  0.507455  ETH  1000.0  USDC  ETHUSDC   

                                           routeix  nroutes        price  \
0                                               20        1  1970.619815   
0                                               22        1  1970.619815   
0                                               24        1  1970.619815   
0                                                0        1  1970.619815   
0                                                2        1  1970.619815   
0                                                4        1  1970.619815   
0                                                6        1  1970.619815   
0                                                8        1  1970.619815   
0                                               10        1  1970.619815   
0                                               12        1  1970.619815   
0                                               14        1  1970.619815   
0                                               16        1  1970.619815   
0                                               18        1  1970.619815   
0  [20, 22, 24, 0, 2, 4, 6, 8, 10, 12, 14, 16, 18]       13  1970.619815   

         p_unit  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH

Then we can set a threshold on the number of orders using the AlphaSim

In [6]:
AlphaSim.amm_buys('USDC',usdc_trade_amount, execute=False, threshold_orders=4)['trades']  # route_trade_by_source

uid  id subid                          note   aggr   exec limitfail  \
0  0.0   0     0                     route #20  False  False      None   
0  0.1   0     1                     route #22  False  False      None   
0  0.2   0     2                     route #24  False  False      None   
0  0.3   0     3                      route #0  False  False      None   
0    0   0     A  AMM sells 1ETH buys 1000USDC   True  False      None   

       amt1 tkn1    amt2  tkn2     pair          routeix  nroutes  \
0  0.100000  ETH   195.0  USDC  ETHUSDC               20        1   
0  0.100000  ETH   195.0  USDC  ETHUSDC               22        1   
0  0.100000  ETH   195.0  USDC  ETHUSDC               24        1   
0  0.207046  ETH   415.0  USDC  ETHUSDC                0        1   
0  0.507046  ETH  1000.0  USDC  ETHUSDC  [20, 22, 24, 0]        4   

         price        p_unit  threshold_orders  
0  1972.205951  USDC per ETH                 4  
0  1972.205951  USDC per ETH                 4  
0  1972.205951  USDC per ETH                 4  
0  1972.205951  USDC per ETH                 4  
0  1972.205951  USDC per ETH                 4

We would expect that routing through fewer orders should give a slightly worse price for the trader.

Since here the amm is buying then the trader is selling 1000 USDC, then the Alpha buy amount of 0.499122 is slightly lower than the 0.49973 ETH achieved in exact routing. However, the difference here is only 7bp

In [7]:
AlphaSim.state()["trades"].query("uid=='0'").iloc[0]["amt1"],Sim.state()["trades"].query("uid=='0'").iloc[0]["amt1"]

(0.507046, 0.507455)

In [8]:
10000*(AlphaSim.state()["trades"].query("uid=='0'").iloc[0]["amt1"]
       /Sim.state()["trades"].query("uid=='0'").iloc[0]["amt1"]-1)

-8.059827965041233

#### AMM buys ETH

In [9]:
Sim.amm_buys('ETH',eth_trade_amount, execute=False)['trades']  # route_trade_by_source

uid  id subid                          note   aggr   exec limitfail  \
0  1.0   1     0                      route #1  False  False      None   
0  1.1   1     1                      route #3  False  False      None   
0  1.2   1     2                      route #5  False  False      None   
0  1.3   1     3                      route #7  False  False      None   
0  1.4   1     4                      route #9  False  False      None   
0  1.5   1     5                     route #11  False  False      None   
0  1.6   1     6                     route #13  False  False      None   
0  1.7   1     7                     route #15  False  False      None   
0  1.8   1     8                     route #17  False  False      None   
0  1.9   1     9                     route #19  False  False      None   
0    1   1     A  AMM sells 1396USDC buys 0ETH   True  False      None   

          amt1  tkn1  amt2 tkn2     pair                              routeix  \
0   139.647707  USDC  0.05  ETH  ETHUSDC                                    1   
0   139.647707  USDC  0.05  ETH  ETHUSDC                                    3   
0   139.647707  USDC  0.05  ETH  ETHUSDC                                    5   
0   139.647707  USDC  0.05  ETH  ETHUSDC                                    7   
0   139.647707  USDC  0.05  ETH  ETHUSDC                                    9   
0   139.647707  USDC  0.05  ETH  ETHUSDC                                   11   
0   139.647707  USDC  0.05  ETH  ETHUSDC                                   13   
0   139.647707  USDC  0.05  ETH  ETHUSDC                                   15   
0   139.647707  USDC  0.05  ETH  ETHUSDC                                   17   
0   139.647707  USDC  0.05  ETH  ETHUSDC                                   19   
0  1396.477067  USDC  0.50  ETH  ETHUSDC  [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]   

   nroutes        price        p_unit  
0        1  2792.954133  USDC per ETH  
0        1  2792.954133  USDC per ETH  
0        1  2792.954133  USDC per ETH  
0        1  2792.954133  USDC per ETH  
0        1  2792.954133  USDC per ETH  
0        1  2792.954133  USDC per ETH  
0        1  2792.954133  USDC per ETH  
0        1  2792.954133  USDC per ETH  
0        1  2792.954133  USDC per ETH  
0        1  2792.954133  USDC per ETH  
0       10  2792.954133  USDC per ETH

In [10]:
AlphaSim.amm_buys('ETH',eth_trade_amount, execute=False, threshold_orders=3)['trades']  # route_trade_by_source

uid  id subid                          note   aggr   exec limitfail  \
0  1.0   1     0                     route #25  False  False      None   
0  1.1   1     1                      route #1  False  False      None   
0  1.2   1     2                      route #3  False  False      None   
0    1   1     A  AMM sells 1383USDC buys 0ETH   True  False      None   

          amt1  tkn1      amt2 tkn2     pair     routeix  nroutes  \
0     1.000000  USDC  0.000364  ETH  ETHUSDC          25        1   
0   690.783942  USDC  0.249818  ETH  ETHUSDC           1        1   
0   690.783942  USDC  0.249818  ETH  ETHUSDC           3        1   
0  1382.567884  USDC  0.500000  ETH  ETHUSDC  [25, 1, 3]        3   

         price        p_unit  threshold_orders  
0  2765.135768  USDC per ETH                 3  
0  2765.135768  USDC per ETH                 3  
0  2765.135768  USDC per ETH                 3  
0  2765.135768  USDC per ETH                 3

Again, we expect that routing through fewer orders should give a slightly worse price for the trader.

Since here the amm is buying then the trader is selling 0.5 ETH, then the Alpha price of 1388.32 USDC means that the return on 0.5 ETH sold is 1388.32 USDC. By comparison, the standard route gives a return of 1392.97 USDC which is slightly better for the trader (33bp)

In [11]:
(AlphaSim.state()["trades"].query("uid=='1'").iloc[0]["amt1"],
 Sim.state()["trades"].query("uid=='1'").iloc[0]["amt1"])

(1382.567884, 1396.477067)

In [12]:
10000*(AlphaSim.state()["trades"].query("uid=='1'").iloc[0]["amt1"]
       /Sim.state()["trades"].query("uid=='1'").iloc[0]["amt1"]-1)

-99.60194355271845

## Route by Target

#### AMM sells USDC

First we test the standard simulator which should route through all 5 positions

In [13]:
Sim.amm_sells('USDC',usdc_trade_amount, execute=False)['trades']  # route_trade_by_target

uid  id subid                          note   aggr   exec limitfail  \
0  2.0   2     0                      route #1  False  False      None   
0  2.1   2     1                      route #3  False  False      None   
0  2.2   2     2                      route #5  False  False      None   
0  2.3   2     3                      route #7  False  False      None   
0  2.4   2     4                      route #9  False  False      None   
0  2.5   2     5                     route #11  False  False      None   
0  2.6   2     6                     route #13  False  False      None   
0  2.7   2     7                     route #15  False  False      None   
0  2.8   2     8                     route #17  False  False      None   
0  2.9   2     9                     route #19  False  False      None   
0    2   2     A  AMM sells 1000USDC buys 0ETH   True  False      None   

     amt1  tkn1      amt2 tkn2     pair                              routeix  \
0   100.0  USDC  0.035779  ETH  ETHUSDC                                    1   
0   100.0  USDC  0.035779  ETH  ETHUSDC                                    3   
0   100.0  USDC  0.035779  ETH  ETHUSDC                                    5   
0   100.0  USDC  0.035779  ETH  ETHUSDC                                    7   
0   100.0  USDC  0.035779  ETH  ETHUSDC                                    9   
0   100.0  USDC  0.035779  ETH  ETHUSDC                                   11   
0   100.0  USDC  0.035779  ETH  ETHUSDC                                   13   
0   100.0  USDC  0.035779  ETH  ETHUSDC                                   15   
0   100.0  USDC  0.035779  ETH  ETHUSDC                                   17   
0   100.0  USDC  0.035779  ETH  ETHUSDC                                   19   
0  1000.0  USDC  0.357788  ETH  ETHUSDC  [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]   

   nroutes        price        p_unit  
0        1  2794.954542  USDC per ETH  
0        1  2794.954542  USDC per ETH  
0        1  2794.954542  USDC per ETH  
0        1  2794.954542  USDC per ETH  
0        1  2794.954542  USDC per ETH  
0        1  2794.954542  USDC per ETH  
0        1  2794.954542  USDC per ETH  
0        1  2794.954542  USDC per ETH  
0        1  2794.954542  USDC per ETH  
0        1  2794.954542  USDC per ETH  
0       10  2794.954542  USDC per ETH

Then we can set a threshold on the number of orders using the AlphaSim

In [14]:
AlphaSim.amm_sells('USDC',usdc_trade_amount, execute=False, threshold_orders = 3)['trades'] # route_trade_by_target

uid  id subid                          note   aggr   exec limitfail  \
0  2.0   2     0                      route #1  False  False      None   
0  2.1   2     1                      route #3  False  False      None   
0  2.2   2     2                      route #5  False  False      None   
0    2   2     A  AMM sells 1000USDC buys 0ETH   True  False      None   

          amt1  tkn1      amt2 tkn2     pair    routeix  nroutes        price  \
0   333.333333  USDC  0.119767  ETH  ETHUSDC          1        1  2783.181806   
0   333.333333  USDC  0.119767  ETH  ETHUSDC          3        1  2783.181806   
0   333.333333  USDC  0.119767  ETH  ETHUSDC          5        1  2783.181806   
0  1000.000000  USDC  0.359301  ETH  ETHUSDC  [1, 3, 5]        3  2783.181806   

         p_unit  threshold_orders  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 3

We would expect that routing through fewer orders should give a slightly worse price for the trader.

Since here the amm is selling then the trader is buying 1000 USDC, then the Alpha price of 0.359301 ETH means the trader is required to pay more ETH for the same 1000 USDC than the standard approach (0.358435 ETH).

This means that the higher ETH amount is a worse price as this is the amount that the trader will part with for a fixed target USDC amount (24bp)

In [15]:
(AlphaSim.state()["trades"].query("uid=='2'").iloc[0]["amt2"],
 Sim.state()["trades"].query("uid=='2'").iloc[0]["amt2"])

(0.359301, 0.357788)

In [16]:
10000*(AlphaSim.state()["trades"].query("uid=='2'").iloc[0]["amt2"]
       /Sim.state()["trades"].query("uid=='2'").iloc[0]["amt2"]-1)

42.28761165830042

#### AMM sells ETH

In [17]:
Sim.amm_sells('ETH',eth_trade_amount, execute=False)['trades']  # route_trade_by_target

uid  id subid                         note   aggr   exec limitfail  amt1  \
0   3.0   3     0                    route #20  False  False      None  0.10   
0   3.1   3     1                    route #22  False  False      None  0.10   
0   3.2   3     2                    route #24  False  False      None  0.10   
0   3.3   3     3                     route #0  False  False      None  0.02   
0   3.4   3     4                     route #2  False  False      None  0.02   
0   3.5   3     5                     route #4  False  False      None  0.02   
0   3.6   3     6                     route #6  False  False      None  0.02   
0   3.7   3     7                     route #8  False  False      None  0.02   
0   3.8   3     8                    route #10  False  False      None  0.02   
0   3.9   3     9                    route #12  False  False      None  0.02   
0  3.10   3    10                    route #14  False  False      None  0.02   
0  3.11   3    11                    route #16  False  False      None  0.02   
0  3.12   3    12                    route #18  False  False      None  0.02   
0     3   3     A  AMM sells 0ETH buys 985USDC   True  False      None  0.50   

  tkn1        amt2  tkn2     pair  \
0  ETH  195.000000  USDC  ETHUSDC   
0  ETH  195.000000  USDC  ETHUSDC   
0  ETH  195.000000  USDC  ETHUSDC   
0  ETH   40.008448  USDC  ETHUSDC   
0  ETH   40.008448  USDC  ETHUSDC   
0  ETH   40.008448  USDC  ETHUSDC   
0  ETH   40.008448  USDC  ETHUSDC   
0  ETH   40.008448  USDC  ETHUSDC   
0  ETH   40.008448  USDC  ETHUSDC   
0  ETH   40.008448  USDC  ETHUSDC   
0  ETH   40.008448  USDC  ETHUSDC   
0  ETH   40.008448  USDC  ETHUSDC   
0  ETH   40.008448  USDC  ETHUSDC   
0  ETH  985.084476  USDC  ETHUSDC   

                                           routeix  nroutes        price  \
0                                               20        1  1970.168952   
0                                               22        1  1970.168952   
0                                               24        1  1970.168952   
0                                                0        1  1970.168952   
0                                                2        1  1970.168952   
0                                                4        1  1970.168952   
0                                                6        1  1970.168952   
0                                                8        1  1970.168952   
0                                               10        1  1970.168952   
0                                               12        1  1970.168952   
0                                               14        1  1970.168952   
0                                               16        1  1970.168952   
0                                               18        1  1970.168952   
0  [20, 22, 24, 0, 2, 4, 6, 8, 10, 12, 14, 16, 18]       13  1970.168952   

         p_unit  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH  
0  USDC per ETH

In [18]:
AlphaSim.amm_sells('ETH',eth_trade_amount, execute=False, threshold_orders = 3)['trades'] # route_trade_by_target

uid  id subid                         note   aggr   exec limitfail  amt1  \
0  3.0   3     0                    route #22  False  False      None   0.1   
0  3.1   3     1                    route #24  False  False      None   0.1   
0  3.2   3     2                     route #0  False  False      None   0.3   
0    3   3     A  AMM sells 0ETH buys 992USDC   True  False      None   0.5   

  tkn1        amt2  tkn2     pair      routeix  nroutes        price  \
0  ETH  195.000000  USDC  ETHUSDC           22        1  1983.812697   
0  ETH  195.000000  USDC  ETHUSDC           24        1  1983.812697   
0  ETH  601.906348  USDC  ETHUSDC            0        1  1983.812697   
0  ETH  991.906348  USDC  ETHUSDC  [22, 24, 0]        3  1983.812697   

         p_unit  threshold_orders  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 3

Again, we expect that routing through fewer orders should give a slightly worse price for the trader.

Since here the amm is selling then the trader is buying 0.5 ETH, then the Alpha price of 1001.762648 USDC means the trader is required to pay more USDC for the same 0.5 ETH than the standard approach (1001.056844 USDC).

This means that the higher USDC amount is a worse price as this is the amount that the trader will part with for a fixed target ETH amount (7bp)

In [19]:
(AlphaSim.state()["trades"].query("uid=='3'").iloc[0]["amt2"],
 Sim.state()["trades"].query("uid=='3'").iloc[0]["amt2"])

(991.906348, 985.084476)

In [20]:
10000*(AlphaSim.state()["trades"].query("uid=='3'").iloc[0]["amt2"]
       /Sim.state()["trades"].query("uid=='3'").iloc[0]["amt2"]-1)

69.25164456657251

### Decreasing the threshold number of orders

It follows that if we decrease the threshold then the price should get worse and vice-versa

In [21]:
AlphaSim.amm_sells('ETH',eth_trade_amount, execute=False, threshold_orders = 2)['trades'] # route_trade_by_target

uid  id subid                         note   aggr   exec limitfail  amt1  \
0  4.0   4     0                    route #22  False  False      None   0.1   
0  4.1   4     1                     route #0  False  False      None   0.4   
0    4   4     A  AMM sells 0ETH buys 998USDC   True  False      None   0.5   

  tkn1        amt2  tkn2     pair  routeix  nroutes        price  \
0  ETH  195.000000  USDC  ETHUSDC       22        1  1996.785314   
0  ETH  803.392657  USDC  ETHUSDC        0        1  1996.785314   
0  ETH  998.392657  USDC  ETHUSDC  [22, 0]        2  1996.785314   

         p_unit  threshold_orders  
0  USDC per ETH                 2  
0  USDC per ETH                 2  
0  USDC per ETH                 2

In [22]:
AlphaSim.amm_sells('ETH',eth_trade_amount, execute=False, threshold_orders = 4)['trades'] # route_trade_by_target

uid  id subid                         note   aggr   exec limitfail  amt1  \
0  5.0   5     0                    route #20  False  False      None   0.1   
0  5.1   5     1                    route #22  False  False      None   0.1   
0  5.2   5     2                    route #24  False  False      None   0.1   
0  5.3   5     3                     route #0  False  False      None   0.2   
0    5   5     A  AMM sells 0ETH buys 986USDC   True  False      None   0.5   

  tkn1       amt2  tkn2     pair          routeix  nroutes        price  \
0  ETH  195.00000  USDC  ETHUSDC               20        1  1971.692739   
0  ETH  195.00000  USDC  ETHUSDC               22        1  1971.692739   
0  ETH  195.00000  USDC  ETHUSDC               24        1  1971.692739   
0  ETH  400.84637  USDC  ETHUSDC                0        1  1971.692739   
0  ETH  985.84637  USDC  ETHUSDC  [20, 22, 24, 0]        4  1971.692739   

         p_unit  threshold_orders  
0  USDC per ETH                 4  
0  USDC per ETH                 4  
0  USDC per ETH                 4  
0  USDC per ETH                 4  
0  USDC per ETH                 4

And indeed reducing to 2 orders gives a worse price (i.e. costs more) and increasing to 4 orders gives a better price (i.e. costs less; 13bp)

In [23]:
(AlphaSim.state()["trades"].query("uid=='4'").iloc[0]["amt2"],
 AlphaSim.state()["trades"].query("uid=='5'").iloc[0]["amt2"])

(998.392657, 985.84637)

In [24]:
10000*(AlphaSim.state()["trades"].query("uid=='4'").iloc[0]["amt2"]
       / AlphaSim.state()["trades"].query("uid=='5'").iloc[0]["amt2"]-1)

127.2641192562296